spec/index.md

```diff
@@ -127,6 +127,10 @@ This specification may refer to certain constants and limits without specifying
 
     Maximum wall clock time spent on evaluation of a query call.
 
+-   `MAX_CALL_TIMEOUT`
+
+    The maximum timeout (in seconds) for an inter-canister call.
```

Comment:
What is the reason we are introducing MAX_CALL_TIMEOUT?
I believe that implicitly bounding timeout can be problematic from a developer experience standpoint. Even more, since the current implementation of the message broadcast is AFAIK equivalent to having an infinite timeout, why we are disallowing that for best-effort messages?

Reply:
What do you think about returning an error if the timeout is larger than `MAX_CALL_TIMEOUT`, instead of implicitly bounding it?

Reply:
TBH, If I were a developer, and if I set a timeout on X seconds, and I received `SYS_UNKNOWN` after X/10 seconds (=`MAX_CALL_TIMEOUT`) I do not believe the first place I would look for a mistake would be interface spec of call_with_best_effort_response.

Reply:
We could return an error on larger timeouts (IMO it would only make sense to return an error on `call_perform` rather than `call_with_best_effort_response`). But, for load shedding, we also want to allow the system to respond with a `SYS_UNKNOWN` even before the timeout elapses, no matter what the timeout is (even if it's below `MAX_CALL_TIMEOUT`). So the canister authors will have to deal with the case of a `SYS_UNKNOWN` being returned sooner, regardless of what we do with bounding the timeout parameter. This obviously needs to be documented well, not just in the spec, but I don't see it changing, otherwise the replica can't be elastic in how it handles load. Given this behavior, we opted not to return an error on exceeding `MAX_CALL_TIMEOUT`.

Another option was to add support for larger timeouts on a best-effort basis, i.e., we accept any timeout, but we reserve the right to invoke a callback earlier with SYS_UNKNOWN if we're under load (which we'd do anyways). But IMO that would complicate the reasoning (1) about keeping the replica state bounded (for example, we plan to support 10s of millions of messages, so we need to make sure that callbacks etc. don't take too much space) while (2) being fair among the canisters (e.g., if one canister has 9 million open calls, we should shed them even if their timeout is years in the feature). And I don't think it changes the situation much for the developers, since they have to deal with the "early reject" anyways.

Reply:
I would like to point out that we have seen such decisions to come back and bite us -- essentially we build something that technically works fine and is well documented but people are surprised by unexpected behaviour or trapped in situations where they cannot differentiate between clearly different error cases that warrant different handling.

I think returning an error when passing what the system considers an invalid parameter (i.e. timeout larger than max allowed) is clearly something different than getting `SYS_UNKNOWN` because the system had to load shed. And people might want to know the difference between the two (similar situation we have with the reject codes where we have people come back to ask how can they differentiate between errors that clearly need different handling). In this case, the former (passing invalid parameter) can be clearly fixed by using an accepted timeout, the latter might well be outside your control (so not much you can do).

Reply:
I'm not dead set against adding an error, but right now I don't think it's the better option.
1. You bring up distinguishing the two relevant error cases from the user perspective. I do actually agree that this is desirable, but I don't see how returning an error on overly large timeouts would solve this robustly. I think the only robust way to distinguish is by looking at the error message (a much better way, as we recently discussed, would be error codes, but that's another topic). Otherwise, what could the user do? They could store the expected deadline when issuing a request, and compare the time of the `SYS_UNKNOWN` reject with the deadline. But this is not robust. First, it could be that the `SYS_UNKNOWN` reject is issued before the deadline because of high load, but stuck in the input queue for a long time. So the user would wrongly conclude that the reject is due to timing out. Conversely (though less likely), with a time drift of even a few seconds, a remote subnet could reject the message early (compared to the local subnet's clock) with a timeout; unless we implement some more complicated machinery, the user could receive this reject before the deadline, breaking their logic again.
2. We have no established way for canisters to query any of the system limits, so I suppose the hard limit on timeouts would be no exception. So the user would have to again dig through the doc to find what the current limit is and hard code it in their canister.
3. In general, but especially given the previous point, we could never lower the max timeout without breaking backwards compatibility. Now sure, the hope is that we designed the whole thing well enough that it stays robust with a 5 minute max timeout, but it seems prudent to be careful.

Reply:
I was thinking that the error when setting a higher than max timeout would be synchronous vs the asynchronous cases you mention (it's kinda similar to my output queue is full vs the receiving canister's input queue is full). However, I now remembered that actually users cannot differentiate between synchronous and asynchronous errors (which yet another topic in itself) so indeed there's probably not much the user can do and as such let's stick to what we have.

Maybe if we get traction on the other effort of distinguishing error codes we can revisit this one too.

Reply:
Since the actual value of `MAX_CALL_TIMEOUT` is not specified, I could see `call_with_best_effort_response()` returning the actual, clamped timeout to give canister developers a hint of the actual outcome.

But returning an error and forcing everyone to deal with it seems to me like a lot more trouble than it's worth. Especially since, as Ogi pointed out, without a specified upper bound we might be breaking canisters if we adjust the bound downward.

In my mind, best-effort calls are just that: best-effort. Anything can go wrong and even if the system makes a good faith effort to differentiate between e.g. locally rejected requests (which are known to not have been delivered, so `SYS_TRANSIENT`) and timed out or shed ones, there is no guarantee that you'll ever get a vague `SYS_UNKNOWN` for your troubles.

Looking at this specific issue, even if the canister can reliably differentiate between a time out and an unceremoniously dropped request, there is no way for it to reliably act on it. For now, when we shed load, we simply drop the largest message. So the canister "dealing with it" might mean breaking down the payload into smaller chunks (assuming that's even a possibility). But in the future we may entirely change the load shedding logic and say "we prefer dropping messages that would be routed into heavily backlogged streams"; or whatever else. There is no way for a canister to deal with all the possibilities.

So I'm thinking we'll do a good faith attempt to differentiate between one and the other, but it's mostly going to be useful for metrics/stats ("X% of my calls time out and Y% are shed, so maybe I should focus on better sharding") as opposed to  "this specific call failed because it was load shed, let me try to resubmit it in smaller chunks".

Reply:
Returning the accepted timeout sounds like a good idea. I don't immediately see how the Rust CDK/Motoko would export this given their current call return types, but I also don't really see anything speaking against it. There's still potential for confusion with interpreting the result (with all the previous caveats applying) but no worse than the current proposal. Wdyt @dragoljub-duric @dsarlis ? 

Re: error handling, I was thinking that the difference in error handling load shedding rejects vs timeout rejects could be "I'll back off from sending messages for a while until the system load subsides" vs "let's either resubmit or query the target whether the message arrived". But yeah, it might be more of a stretch, especially given the current lack of a `sleep` call.